### ISSUE TITLE
[🐛 Bug]: Selenium not initializing properly in container

### ISSUE THREAD

**Issue by 19peytonsmith** [SeleniumHQ/docker-selenium#1969]
### What happened?

Everytime the server starts both locally and on dev server, the browser container outputs errors initially, and when using selenium throughout the code I get errors as a result.

Specifically, calling Selenium::WebDriver.for(:remote, url: ENV['SELENIUM_REMOTE_URL'], capabilities: caps) yields these error logs:
![image](https://github.com/SeleniumHQ/docker-selenium/assets/54728307/98b468fd-bde1-40ae-9884-7e6198e94030)

A few months ago this feature was working, haven't tested it since and it looks like it broke in the meantime, not sure how, thinking maybe it's a dependency version issue.

### Command used to start Selenium Grid with Docker

```shell
version: '3.9'
services:
  browser:
    image: selenium/standalone-chrome
    ports:
      - '4444:4444'
    volumes:
      - /dev/shm:/dev/shm
    restart: always
```

### Relevant log output

```shell
2023-10-09 13:55:00 17:55:00.758 WARN [DriverServiceSessionFactory.apply] - Error while creating session with the driver service. Stopping driver service: Could not start a new session. Response code 500. Message: session not created: Chrome failed to start: exited normally.
2023-10-09 13:55:00   (session not created: DevToolsActivePort file doesn't exist)
2023-10-09 13:55:00   (The process started from chrome location /opt/google/chrome/chrome is no longer running, so ChromeDriver is assuming that Chrome has crashed.) 
2023-10-09 13:55:00 Host info: host: '7bf59e95670c', ip: '172.18.0.3'
2023-10-09 13:55:00 Build info: version: '4.12.1', revision: '8e34639b11'
2023-10-09 13:55:00 System info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.10.16.3-microsoft-standard-WSL2', java.version: '11.0.20.1'
2023-10-09 13:55:00 Driver info: driver.version: unknown
2023-10-09 13:55:00 17:55:00.779 WARN [SeleniumSpanExporter$1.lambda$export$1] - Error while creating session with the driver service. Stopping driver service: Could not start a new session. Response code 500. Message: session not created: Chrome failed to start: exited normally.
2023-10-09 13:55:00   (session not created: DevToolsActivePort file doesn't exist)
2023-10-09 13:55:00   (The process started from chrome location /opt/google/chrome/chrome is no longer running, so ChromeDriver is assuming that Chrome has crashed.) 
2023-10-09 13:55:00 Host info: host: '7bf59e95670c', ip: '172.18.0.3'
2023-10-09 13:55:00 Build info: version: '4.12.1', revision: '8e34639b11'
2023-10-09 13:55:00 System info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.10.16.3-microsoft-standard-WSL2', java.version: '11.0.20.1'
2023-10-09 13:55:00 Driver info: driver.version: unknown
2023-10-09 13:55:00 17:55:00.780 WARN [SeleniumSpanExporter$1.lambda$export$1] - org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Response code 500. Message: session not created: Chrome failed to start: exited normally.
2023-10-09 13:55:00   (session not created: DevToolsActivePort file doesn't exist)
2023-10-09 13:55:00   (The process started from chrome location /opt/google/chrome/chrome is no longer running, so ChromeDriver is assuming that Chrome has crashed.) 
2023-10-09 13:55:00 Host info: host: '7bf59e95670c', ip: '172.18.0.3'
2023-10-09 13:55:00 Build info: version: '4.12.1', revision: '8e34639b11'
2023-10-09 13:55:00 System info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.10.16.3-microsoft-standard-WSL2', java.version: '11.0.20.1'
2023-10-09 13:55:00 Driver info: driver.version: unknown
2023-10-09 13:55:00     at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:140)
2023-10-09 13:55:00     at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:96)
2023-10-09 13:55:00     at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:68)
2023-10-09 13:55:00     at org.openqa.selenium.grid.node.config.DriverServiceSessionFactory.apply(DriverServiceSessionFactory.java:167)
2023-10-09 13:55:00     at org.openqa.selenium.grid.node.config.DriverServiceSessionFactory.apply(DriverServiceSessionFactory.java:73)
2023-10-09 13:55:00     at org.openqa.selenium.grid.node.local.SessionSlot.apply(SessionSlot.java:147)
2023-10-09 13:55:00     at org.openqa.selenium.grid.node.local.LocalNode.newSession(LocalNode.java:468)
2023-10-09 13:55:00     at org.openqa.selenium.grid.distributor.local.LocalDistributor.startSession(LocalDistributor.java:648)
2023-10-09 13:55:00     at org.openqa.selenium.grid.distributor.local.LocalDistributor.newSession(LocalDistributor.java:565)
2023-10-09 13:55:00     at org.openqa.selenium.grid.distributor.local.LocalDistributor$NewSessionRunnable.handleNewSessionRequest(LocalDistributor.java:829)
2023-10-09 13:55:00     at org.openqa.selenium.grid.distributor.local.LocalDistributor$NewSessionRunnable.lambda$run$1(LocalDistributor.java:785)
2023-10-09 13:55:00     at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2023-10-09 13:55:00     at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2023-10-09 13:55:00     at java.base/java.lang.Thread.run(Thread.java:829)
2023-10-09 13:55:00 
2023-10-09 13:55:00 17:55:00.782 WARN [SeleniumSpanExporter$1.lambda$export$3] - {"traceId": "3b0b579a6691c9e77fe376a669ccf3b7","eventTime": 1696874100758042200,"eventName": "exception","attributes": {"driver.url": "http:\u002f\u002flocalhost:4985","exception.message": "Error while creating session with the driver service. Stopping driver service: Could not start a new session. Response code 500. Message: session not created: Chrome failed to start: exited normally.\n  (session not created: DevToolsActivePort file doesn't exist)\n  (The process started from chrome location \u002fopt\u002fgoogle\u002fchrome\u002fchrome is no longer running, so ChromeDriver is assuming that Chrome has crashed.) \nHost info: host: '7bf59e95670c', ip: '172.18.0.3'\nBuild info: version: '4.12.1', revision: '8e34639b11'\nSystem info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.10.16.3-microsoft-standard-WSL2', java.version: '11.0.20.1'\nDriver info: driver.version: unknown","exception.stacktrace": "org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Response code 500. Message: session not created: Chrome failed to start: exited normally.\n  (session not created: DevToolsActivePort file doesn't exist)\n  (The process started from chrome location \u002fopt\u002fgoogle\u002fchrome\u002fchrome is no longer running, so ChromeDriver is assuming that Chrome has crashed.) \nHost info: host: '7bf59e95670c', ip: '172.18.0.3'\nBuild info: version: '4.12.1', revision: '8e34639b11'\nSystem info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.10.16.3-microsoft-standard-WSL2', java.version: '11.0.20.1'\nDriver info: driver.version: unknown\n\tat org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:140)\n\tat org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:96)\n\tat org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:68)\n\tat org.openqa.selenium.grid.node.config.DriverServiceSessionFactory.apply(DriverServiceSessionFactory.java:167)\n\tat org.openqa.selenium.grid.node.config.DriverServiceSessionFactory.apply(DriverServiceSessionFactory.java:73)\n\tat org.openqa.selenium.grid.node.local.SessionSlot.apply(SessionSlot.java:147)\n\tat org.openqa.selenium.grid.node.local.LocalNode.newSession(LocalNode.java:468)\n\tat org.openqa.selenium.grid.distributor.local.LocalDistributor.startSession(LocalDistributor.java:648)\n\tat org.openqa.selenium.grid.distributor.local.LocalDistributor.newSession(LocalDistributor.java:565)\n\tat org.openqa.selenium.grid.distributor.local.LocalDistributor$NewSessionRunnable.handleNewSessionRequest(LocalDistributor.java:829)\n\tat org.openqa.selenium.grid.distributor.local.LocalDistributor$NewSessionRunnable.lambda$run$1(LocalDistributor.java:785)\n\tat java.base\u002fjava.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base\u002fjava.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat java.base\u002fjava.lang.Thread.run(Thread.java:829)\n","exception.type": "org.openqa.selenium.SessionNotCreatedException","logger": "org.openqa.selenium.grid.node.config.DriverServiceSessionFactory","session.capabilities": "{\"browserName\": \"chrome\",\"browserVersion\": \"117.0\",\"platformName\": \"linux\",\"se:noVncPort\": 7900,\"se:vncEnabled\": true}\n"}}
```

### Operating System

5.10.16.3-microsoft-standard-WSL2

### Docker Selenium version (tag)

docker.io/selenium/standalone-chrome:117.0

**Comment by 19peytonsmith** [comment]
In my gemfile I have these gems associated:
```
gem 'selenium-webdriver'
gem 'webdrivers', require: !ENV['SELENIUM_REMOTE_URL']
```

in my gemfile.lock here are the versions:
```
    selenium-webdriver (4.2.1)
      childprocess (>= 0.5, < 5.0)
      rexml (~> 3.2, >= 3.2.5)
      rubyzip (>= 1.2.2, < 3.0)
      websocket (~> 1.0)
    webdrivers (5.0.0)
      nokogiri (~> 1.6)
      rubyzip (>= 1.3.0)
      selenium-webdriver (~> 4.0)
```

SELENIUM_REMOTE_URL is an env var set to `http://browser:4444/wd/hub`

Honestly that's all the bit of code that I can provide & is relevant toward getting selenium up and running, I provided the docker-compose service that I am using above. When I go to `http://localhost:4444/` it shows the selenium grid UI. Not sure much else that I can tell you.

**Comment by diemol** [comment]
We do not maintain the `webdrivers` gem, and I don't why you need it if you are using a Grid.

**Comment by 19peytonsmith** [comment]
Okay, removed webdrivers gem and upgraded selenium-webdrivers to 4.14.0. Here I implemented the test example as per the selenium.dev docs

```
require 'selenium-webdriver'
class RecordSpeedtestJob
  include Sidekiq::Worker

  def perform(params = {})

    options = Selenium::WebDriver::Options.chrome
    driver = Selenium::WebDriver.for :remote, url: ENV['SELENIUM_REMOTE_URL'], capabilities: options

    driver.get('https://www.selenium.dev/selenium/web/web-form.html')
    
    driver.title
    
    driver.manage.timeouts.implicit_wait = 500
    
    text_box = driver.find_element(name: 'my-text')
    submit_button = driver.find_element(tag_name: 'button')
    
    text_box.send_keys('Selenium')
    submit_button.click
    
    message = driver.find_element(id: 'message')
    driver.quit
  end
end
```

Here's my test, I was able to remove the driver not initializing error, but now when it sends the request to the browser container, I get errors there, here's the error:
`Selenium::WebDriver::Error::SessionNotCreatedError: Could not start a new session. Error while creating session with the driver service. Stopping driver service: Could not start a new session. Response code 500. Message: session not created: Chrome failed to start: exited normally.`. In the browser container I am getting the same logs as above in 'Relevant log output'

**Comment by diemol** [comment]
What is your host operating system and host machine?

**Comment by 19peytonsmith** [comment]
Windows 11 is the OS on my local system, i5 9600k, 32gb RAM. I added `disable-dev-shm-usage` to `Selenium::WebDriver::Chrome::Options.new(options: { args: ... }) ` which got me able to run the job for the first time, but I've been running into timeout issues sometimes, other times it would work. I also removed the mounting of /dev/shm from my docker compose. So yeah, all in all, for my system I am using windows docker and it doesn't support /dev/shm properly, performance is unreliable but it's manageable for testing.
